cmds/bootloader.c

```diff
@@ -0,0 +1,129 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Operating system loader
+ *
+ * enters ROM bootloader mode
+ *
+ * Copyright 2023 Phoenix Systems
+ * Author: Gerard Swiderski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include "cmd.h"
+
+#include <devices/devs.h>
+#include <hal/hal.h>
+#include <hal/armv7m/imxrt/bootloader.h>
+#include <lib/lib.h>
+
+
+static void cmd_bootloaderInfo(void)
+{
+	lib_printf("enters cpu vendor bootloader mode use <-h> to see usage");
+}
+
+
+static void print_usage(const char *name)
+{
+	lib_printf(
+		"Usage: %s <-b | -s imagenum>\n"
+		"where:\n"
+		"\t-b enters USB serial downloader mode\n"
+		"\t-s <imagenum> selects image <1, 2, 3 or 4> to boot from\n",
+		name);
+}
+
+
+static int cmd_bootloader(int argc, char *argv[])
+{
+	int ret;
+	int opt;
+	int optBootMode = 0;
+	int optImageSel = 0;
+
+	lib_printf("\n");
+
+	ret = bootloader_init();
+	if (ret < 0) {
+		log_error("%s: ROM bootloader unavailable", argv[0]);
+		return ret;
+	}
+
+	if (argc == 1) {
+		print_usage(argv[0]);
+		return CMD_EXIT_FAILURE;
+	}
+
+	for (;;) {
+		opt = lib_getopt(argc, argv, "bhs:");
+		if (opt == -1) {
+			break;
+		}
+
+		switch (opt) {
+			case 'b':
+				optBootMode = BOOT_DOWNLOADER_USB;
+				break;
+
+			case 's':
+				optImageSel = (int)lib_strtol(optarg, NULL, 0);
+				if ((optImageSel < 1) || (optImageSel > 4)) {
+					log_error("%s: invalid image selection <1, 4>\n", argv[0]);
+					print_usage(argv[0]);
+					return CMD_EXIT_FAILURE;
+				}
+				break;
+
+			case 'h':
+			default:
+				print_usage(argv[0]);
+				return CMD_EXIT_FAILURE;
+		}
+	}
+
+	if ((optBootMode != 0) && (optImageSel != 0)) {
+		log_error("%s: select either boot mode or image number, not both\n", argv[0]);
+		print_usage(argv[0]);
+		return CMD_EXIT_FAILURE;
+	}
+
+	optImageSel = ((optImageSel < 1) && (optImageSel > 4)) ? 0 : (optImageSel - 1);
```

Comment:
Nitpick - there's no way for `optImageSel` to be `> 4`
```suggestion
	if (optImageSel > 0) {
		--optImageSel;
	}
```

Or, more hacky way:
```
optImageSel = 1;
(...)
--optImageSel;
```

Reply:
Done, thanks.

Reply:
I think you missed `= 1;` part though